src/axom/quest/readers/C2CReader.cpp

```diff
@@ -0,0 +1,372 @@
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#include "axom/quest/readers/C2CReader.hpp"
+
+#ifndef AXOM_USE_C2C
+  #error C2CReader should only be included when Axom is configured with C2C
+#endif
+
+#include "axom/core.hpp"
+#include "axom/slic.hpp"
+#include "axom/primal.hpp"
+#include "fmt/fmt.hpp"
+
+namespace axom
+{
+namespace quest
+{
+/// returns the linear interpolation of \a A and \a B at \a t. i.e. (1-t)A+tB
+inline double lerp(double A, double B, double t) { return (1 - t) * A + t * B; }
+
+/*!
+ * \brief Helper class for interpolating points on a NURBS curve
+ *
+ * This class was adapted from a similar class in the C2C library's testing framework.
+ * The algorithms are adapted from: Piegl and Tiller's "The NURBS Book", 2nd Ed,  (Springer, 1997) 
+ */
+struct NURBSInterpolator
+{
+  using BasisVector = std::vector<double>;
+  using PointType = primal::Point<double, 2>;
+
+  // EPS is used for computing span intervals
+  NURBSInterpolator(const c2c::NURBSData& curve, double EPS = 1E-9)
+    : m_curve(curve)
+  {
+    int p = m_curve.order - 1;
+    int knotSize = m_curve.knots.size();
+    SLIC_ASSERT(p >= 1);
+    SLIC_ASSERT(knotSize >= 2 * p);
+    computeSpanIntervals(EPS);
+  }
+
+  /// Helper function to compute the start and end parametric coordinates of each knot span
+  void computeSpanIntervals(double EPS)
+  {
+    using axom::utilities::isNearlyEqual;
+    const auto& U = m_curve.knots;
+    const int knotSize = U.size();
+    const int p = m_curve.order - 1;
+    const int n = knotSize - 1 - p - 1;
+
+    for(int i = p; i < n + 1; ++i)
+    {
+      const double left = U[i];
+      const double right = U[i + 1];
+      if(!isNearlyEqual(left, right, EPS))
+      {
+        m_spanIntervals.push_back(std::make_pair(left, right));
+      }
+    }
+  }
+
+  /*!
+   * \brief Checks that the knots of the NURBS curve are closed
+   *
+   * The knots vector is closed when it begins with \a m_curve.order equal knot values
+   * and ends with the same number of equal knot values
+   */
+  bool areKnotsClosed(double EPS = 1E-9) const
+  {
+    using axom::utilities::isNearlyEqual;
+    const auto& U = m_curve.knots;
+
+    int p = m_curve.order - 1;
+
+    const double startKnot = U[0];
+    for(int i = 1; i <= p; ++i)
+    {
+      if(!isNearlyEqual(startKnot, U[i], EPS))
+      {
+        return false;
+      }
+    }
+
+    const int endIndex = U.size() - 1;
+    const double endKnot = U[endIndex];
+    for(int i = 1; i <= p; ++i)
+    {
+      const int index = endIndex - p - 1;
+      if(!isNearlyEqual(endKnot, U[index], EPS))
+      {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  int numSpans() const { return m_spanIntervals.size(); }
+
+  double startParameter(int span = -1) const
+  {
+    const bool inRange = span >= 0 && span < numSpans();
+    return inRange ? m_spanIntervals[span].first : m_curve.knots[0];
+  }
+  double endParameter(int span = -1) const
```

Comment:
this is missing a line, did you run this through clangformat?

Reply:
The code was run through clang-format. 
It seems that our rules don't enforce a newline between functions.

I manually added the line.